### ISSUE TITLE
Ordering of overloads in generic classes and Unknown types

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#3252]
For @erictraut     and @gramster 

I'm not sure if this is a bug or not, and it probably illustrates something with type checkers that might not be possible.  Context is some things I am trying to fix in the pandas type stubs that are shipped with VS Code, and I'm stuck.

There are two files included, `genovertst.py` and `genover.pyi` .  Place them in a directory, and use pylance 2022.3.3 with VS Code.

The stubs define a `Series` that is `Generic`, but we want to allow the user of pandas to not specify the full type (e.g., not write `Series[str]`), so pyright will report `Series[Unknown]`, which is OK.  The question is when you have an object that has that type, how overloads are matched.

In the example, there are two methods, `f1` and `f2` .  They have the same signatures, but the only difference is the order of the overloads.

All the results with `reveal_type()` make sense to me, EXCEPT for the result for `s2`, which I would have expected to be `Series[Unknown]` .

So the question is what is the right order of the overloads to use so that people can create a `Series`, but have the `Unknown` aspect of `Series[Unknown]` be carried through by the type checker?  Is the order for `f1` correct?

It may be the case that making `Series` generic in the stubs, while the implementation does not use `Generic` just won't work.  Since @gramster was the one who put the `Generic` stuff in there for the type stubs, I'd like to understand a way forward.

`genovertst.py`:
```python
from genover import Series, Timedelta

td = Timedelta()

s = Series()
reveal_type(s)

std = s.gettd()
reveal_type(std)

s1 = s.f1(4)
reveal_type(s1)

st1 = std.f1(4)
reveal_type(st1)

s1td = s.f1(td)
reveal_type(s1td)

st1td = std.f1(td)
reveal_type(s1td)

s2 = s.f2(4)
reveal_type(s2)

st2 = std.f2(4)
reveal_type(st2)

s2td = s.f2(td)
reveal_type(s2td)

st2td = std.f2(td)
reveal_type(s2td)
```

`genover.pyi`:
```python
from typing import Generic, TypeVar, Union, overload
import numpy as np

class Timedelta: ...

S1 = TypeVar("S1", str, Timedelta)
num = Union[int, float]

class Series(Generic[S1]):
    def __init__(self) -> None: ...
    @overload
    def f1(
        self,
        other: num,
    ) -> Series[S1]: ...
    @overload
    def f1(
        self: Series[Timedelta],
        other: Union[num, Timedelta],
    ) -> Series[Timedelta]: ...
    
    @overload
    def f2(
        self: Series[Timedelta],
        other: Union[num, Timedelta],
    ) -> Series[Timedelta]: ...
    @overload
    def f2(
        self,
        other: num,
    ) -> Series[S1]: ...

    def gettd(self) -> Series[Timedelta]: ...
```

Note that `mypy` 0.931 has an even funkier result - it doesn't carry the `Unknown` and just picks `str` as the "Unknown" type:
```
genovertst.py:6: note: Revealed type is "genover.Series[builtins.str*]"
genovertst.py:9: note: Revealed type is "genover.Series[genover.Timedelta]"
genovertst.py:12: note: Revealed type is "genover.Series[builtins.str*]"
genovertst.py:15: note: Revealed type is "genover.Series[genover.Timedelta*]"
genovertst.py:17: error: Argument 1 to "f1" of "Series" has incompatible type "Timedelta"; expected "Union[int, float]"
genovertst.py:18: note: Revealed type is "genover.Series[builtins.str*]"
genovertst.py:21: note: Revealed type is "genover.Series[builtins.str*]"
genovertst.py:24: note: Revealed type is "genover.Series[builtins.str*]"
genovertst.py:27: note: Revealed type is "genover.Series[genover.Timedelta]"
genovertst.py:29: error: Argument 1 to "f2" of "Series" has incompatible type "Timedelta"; expected "Union[int, float]"
genovertst.py:30: note: Revealed type is "genover.Series[builtins.str*]"
genovertst.py:33: note: Revealed type is "genover.Series[builtins.str*]"
```

**Comment by erictraut** [comment]
Overload matching behavior is not specified in PEP 484, but type checkers have (mostly) standardized in the approach used.

In the case of pyright, we do the following:
1. Compare the argument list against the parameter list and filter out any overloads that don't match in argument number or keyword name. This eliminates any obvious mismatches.
2. Evaluate the types of the arguments and determine if any of them are unions.
3. Walk through the remaining overloads in declaration order and match the argument types against the parameter types. The first match "wins".
4. If none of the overloads matched but one or more of the arguments evaluated to union types, expand those union types as though the caller were invoking the call multiple times with different argument lists. Re-evaluate step 3 with the expanded list of argument types, possibly matching multiple overloads. Union the resulting types together.

In your sample, your overload signatures are distinguished by an annotation on `self`. The type `Series[Unknown]` is compatible with type `Series[Timedelta]`, so the first overload is chosen (as per step 3 above).

There is an active discussion in the typing-sig about adding support for default type arguments. See https://mail.python.org/archives/list/typing-sig@python.org/thread/7VWBZWXTCX6RAJO6GG67BAXUPFZ24NTC/ for details. I think this is the functionality you are looking for in this case. But the proposal is still in the early draft stage and probably won't make it in time for Python 3.11.
